core/src/main/scala/org/apache/spark/sql/streaming/SnappySinkCallback.scala

```diff
@@ -0,0 +1,184 @@
+        case exception: Throwable => throw exception
+      }
+    }
+
+    sinkCallback.process(snappySession, parameters, batchId, convert(data), posDup)
+  }
+
+  /**
+    * This conversion is necessary as Sink
+    * documentation disallows an operation on incoming dataframe.
+    * Otherwise it will break incremental planning of streaming dataframes.
+    * See http://apache-spark-developers-list.1001551.n3.nabble.com/
+    * Structured-Streaming-Sink-in-2-0-collect-foreach-restrictions-added-in-
+    * SPARK-16020-td18118.html
+    * for a detailed discussion.
+    */
+  def convert(ds: DataFrame): DataFrame = {
+    snappySession.internalCreateDataFrame(
+      ds.queryExecution.toRdd,
+      StructType(ds.schema.fields))
+  }
+}
+
+object DefaultSnappySinkCallback {
+  private val log = Logger.getLogger(classOf[DefaultSnappySinkCallback].getName)
+}
+
+import org.apache.spark.sql.snappy._
+
+class DefaultSnappySinkCallback extends SnappySinkCallback {
+  def process(snappySession: SnappySession, parameters: Map[String, String],
+              batchId: Long, df: Dataset[Row], posDup: Boolean) {
+
+    df.cache()
+    val snappyTable = parameters(TABLE_NAME).toUpperCase
+    DefaultSnappySinkCallback.log.debug(s"Processing for $snappyTable and batchId $batchId")
+
+    val tableName = parameters(TABLE_NAME).toUpperCase
+    val keyColumnsDefined = !snappySession.sessionCatalog.getKeyColumns(tableName).head(1).isEmpty
+    val eventTypeColumnAvailable = df.schema.map(_.name).contains(EVENT_TYPE_COLUMN)
+    if (keyColumnsDefined) {
+      if (eventTypeColumnAvailable) {
+        // TODO: handle scenario when a batch contain multiple records with
+        // same key columns
+        val deleteDf = df.filter(df(EVENT_TYPE_COLUMN) === EventType.DELETE)
+          .drop(EVENT_TYPE_COLUMN)
+        deleteDf.write.deleteFrom(tableName)
+        if (posDup) {
+          val upsertEventTypes = List(EventType.INSERT, EventType.UPDATE)
+          val upsertDf = df.filter(df(EVENT_TYPE_COLUMN).isin(upsertEventTypes: _*))
+              .drop(EVENT_TYPE_COLUMN)
+          upsertDf.write.putInto(tableName)
+        } else {
+          val insertDf = df.filter(df(EVENT_TYPE_COLUMN) === EventType.INSERT)
+            .drop(EVENT_TYPE_COLUMN)
+          insertDf.write.insertInto(tableName)
+          val updateDf = df.filter(df(EVENT_TYPE_COLUMN) === EventType.UPDATE)
+            .drop(EVENT_TYPE_COLUMN)
+          updateDf.write.putInto(tableName)
+        }
+      } else {
+        df.write.putInto(tableName)
+      }
+    }
+    else {
+      if (eventTypeColumnAvailable) {
+        val msg = s"$EVENT_TYPE_COLUMN is present in data but key columns are not defined on table."
+        throw new IllegalStateException(msg)
+      } else {
+        df.write.insertInto(tableName)
+      }
+    }
+    // test hook for validating idempotency
+    if (parameters.contains("internal___failBatch")
+        && parameters("internal___failBatch") == "true") {
```

Comment:
Can move this out as a constant with name indicating test hook like TEST_FAILBATCH_OPTION